design/ceph/osd-migration.md

```diff
@@ -38,137 +27,101 @@ considered in the future:
-storage:
-  store:
-    type: bluestore-rdr
-    updateStore: yes-really-update-store
-```
-
-#### Replace OSD
-The operator's reconciler will replace one OSD at a time.
-A configmap will be used to store the OSD ID currently being migrated.
 OSD replacement steps:
-  1. List all OSDs where `osd-store:<osd store type>` does not match `spec.storage.store.type`.
-  2. If all PGs are not `active+clean`, do not proceed.
-  3. If all PGs  are `active+clean` but a previous OSD replacement is not completed, do not proceed.
-  4. If all PGs are `active+clean` and no replacement is in progress, then select an OSD to be migrated.
-  5. Delete the OSD deployment.
-  6. Create an OSD prepare job with an environment variable indictating the OSD ID to be replaced.
-  7. The OSD prepare pod will destroy the OSD and prepare it again using the same OSD ID. Refer [Destroy OSD](#destroy-osd) for details.
-  8. Once the destroyed OSD pod is recreated, delete the configmap.
-  9. If there is any error during the OSD migration, then preserve the OSD ID being replaced in the configmap for next reconcile.
-  10. Reconcile the operator and perform as same steps until all the OSDs have migrated to the new backend.
-
-#### Destroy OSD
-
-The OSD prepare pod job will destroy an OSD using following steps:
-- Check for OSD ID environment variable of the OSD to be destroyed.
-- Use `ceph volume list` to fetch the OSD path.
-- Destroy the OSD using following command:
-
-    ```
-    ceph osd destroy <OSD_ID> --yes-i-really-mean-it
-    ```
-
-- Wipe the OSD drive. This removes all the data on the device.
-- Prepare the OSD with the new store type by using the same OSD ID. This is done by passing the OSD ID as `--osd-id` flag to `ceph-volume` command.
-
-    ```
-    ceph-volume lvm prepare --osd-id <OSD_ID> --data <OSD_PATH>
-    ```
-
-## Planning
-
-These changes require significant development effort to migrate existing OSDs to use a new backend.
-They will be divided into following phases:
-* New OSDs (greenfield).
-* Migrating existing OSDs on PVC without metadata devices.
-* Migrating existing OSDs on PVC with metadata devices.
-* Existing node-based OSDs (multiple OSDs are created at once via `ceph-volume batch` which adds additional complications).
+
+1. List all OSDs where `osd-store:<osd store type>` does not match `spec.storage.store.type`.
+1. If all PGs are not `active+clean`, do not proceed.
+1. If all PGs are `active+clean` but a previous OSD migration is not completed, do not proceed.
+1. If all PGs are `active+clean` and no migration is in progress, then select an OSD to be migrated.
+1. Delete the OSD deployment.
+1. Create an OSD prepare job with an environment variable indictating the OSD ID to be replaced.
+1. The OSD prepare pod will destroy the OSD (`ceph osd destroy {id} --yes-i-really-mean-it`) and prepare it again using the same OSD ID. Refer [Destroy OSD](#destroy-osd) for details.
+1. Once the destroyed OSD pod is recreated, delete the configmap.
```

Comment:
```suggestion
1. Once the destroyed OSD pod is recreated, delete the configmap.
```

question: if there are more than 1 osd to migrate, we'll be creating new confogmap each time?

Reply:
yes, the idea is to delete the configmap after migration of each OSD.

Reply:
If the OSD migration goes wrong, reconciling the operator might not resolve it, as the OSD migraiton requires a number of step.  So if the configMap from previous migration is still present, we can safely assume the last migration didn't succeed. So rook can halt any further migration.

Reply:
This is a really good detail. Thanks Santosh. Since these are long running operations it's much more likely that the operator might have an issue during the process, and we have to make sure Rook can restart close where it left off without issues.